fix(auctions): Closed auctions fixes

pvinis opened the pull request:
The type of this PR is: **Bugfix**

<!-- Bugfix/Feature/Enhancement/Documentation -->

<!-- If applicable, write the Jira ticket number in square brackets e.g. `[CX-434]`
     The Jira integration will turn it into a clickable link for you. -->

This PR resolves [CX-725], [CX-728]

### Description

<!-- Implementation description -->

This includes two fixes:
- when a closed auction is accessed, it doesn't try to go live anymore.
- when a closed auction is accessed, the register button is not displayed.

### PR Checklist (tick all before merging)

<!-- 💡 This checklist is experimental. MX warmly welcomes any feedback about the list or how it impacts your workflow -->

- [x] I have included screenshots or videos to illustrate my changes, or I have not changed anything that impacts the UI.
- [x] I have added tests for my changes, or my changes don't require testing, or I have included a link to a separate Jira ticket covering the tests.
- [x] I have documented any follow-up work that this PR will require, or it does not require any.
- [x] I have added an app state migration, or my changes do not require one. ([What are migrations?](https://github.com/artsy/eigen/blob/master/docs/adding_state_migrations.md))
- [x] I have added a [CHANGELOG.yml](/CHANGELOG.yml) entry or my changes do not require one.


[CX-434]: https://artsyproduct.atlassian.net/browse/CX-434
[CX-725]: https://artsyproduct.atlassian.net/browse/CX-725

[CX-728]: https://artsyproduct.atlassian.net/browse/CX-728

MounirDhahri commented on src/lib/Scenes/Sale/Sale.tsx:
```diff
@@ -48,29 +51,30 @@ export const Sale: React.FC<Props> = ({ queryRes }) => {
+  useInterval(() => {
+    const now = moment()
+    setIsLive(
+      sale.liveStartAt !== null && now.isAfter(sale.liveStartAt) && sale.endAt !== null && now.isBefore(sale.endAt)
```
I think we should also make this check if the sale `endAt` is null and we're after the `liveStartAt`

pvinis replied:
🤔 makes sense.

MounirDhahri commented on src/lib/utils/usePrevious.ts:
```diff
@@ -0,0 +1,11 @@
+
+  useEffect(() => {
+    ref.current = value
+  }, [value])
```
Since we have value in the dependencies array I guess it's no longer the previous value right? it will be updated once the isLive value is updated and the screen is mounted

MounirDhahri replied:
**thought:** I guess you are trying to get the new `isLive` value before the screen is mounted, if that's the case I think you should move the reference and the the `useEffect`to the sale screen

pvinis replied:
Mm I'm not sure what you mean. I used https://usehooks.com/usePrevious/ and modified it to have an initial value. The return is happening before the useEffect, so the usePrevious hook always gives us the previous value. What I use it for is to know when `isLive` is `true` and `prevIsLive` is `false`. That's the only case where we should navigate to the live screen. In any other case we do nothing.

MounirDhahri replied:
Sorry, I must have misunderstood the effect then 😬

MounirDhahri commented:
Thanks for fixing these fast 🙏, I left few minor **non-blocking** comments. But I think we should also add a test for `when a closed auction is accessed, it doesn't try to go live anymore.` to avoid this regression in the future

MounirDhahri commented:
Thanks for fixing these fast 🙏 good job. I left few minor non-blocking comments. But I think we should also add a test for _when a closed auction is accessed, it doesn't try to go live anymore._ to avoid this regression in the future

pvinis commented:
Yup, I'll add that too.

brainbicycle commented on src/lib/Scenes/Sale/__tests__/Sale-tests.tsx:
```diff
@@ -78,22 +71,79 @@ describe("Sale", () => {
+      })
+    )
+
+    expect(tree.findAllByType(RegisterToBidButton)).toHaveLength(0)
```
🎉

brainbicycle commented on src/lib/Scenes/Sale/Sale.tsx:
```diff
@@ -48,29 +51,30 @@ export const Sale: React.FC<Props> = ({ queryRes }) => {
+    setIsLive(
+      sale.liveStartAt !== null && now.isAfter(sale.liveStartAt) && sale.endAt !== null && now.isBefore(sale.endAt)
+    )
+  }, 500)
```
nice new useInterval is cleaner ✨

pvinis commented:
This should be ready. I'll add @admbtlr as a reviewer because I changed stuff in the `saleTime`. I tried to make types strict, and ended up with a couple of edge cases that were missing. I talked with Sam for how to handle these, changed the tests to reflect that, and then changed the code to make them all pass.

admbtlr commented on src/lib/Scenes/Sale/Sale.tsx:
```diff
@@ -8,6 +8,9 @@ import { defaultEnvironment } from "lib/relay/createEnvironment"
 import { Schema } from "lib/utils/track"
+import { useInterval } from "lib/utils/useInterval"
+import { usePrevious } from "lib/utils/usePrevious"
+import _ from "lodash"
```
Shouldn't this rather be 
```
import compact from 'lodash/compact'
```
in order to avoid importing the whole caboodle?

pvinis replied:
Nothing really changes in RN. We are not web (thankfully, in this case 😅). In most files we import like that, and since there is no change in perf or whatever, why not keep it consistent and simple. 👍

admbtlr commented on src/lib/Scenes/Sale/Sale.tsx:
```diff
@@ -48,29 +51,45 @@ export const Sale: React.FC<Props> = ({ queryRes }) => {
+  // poll every .5 seconds to check if sale has gone live
+  useInterval(() => {
+    const now = moment()
+    if (sale.liveStartAt === null) {
```
If it's null, then we don't need to do the `useInterval` in the first place, do we? Unless we're expecting that value to change from null at some point?

pvinis replied:
Indeed. That was one of my thoughts against using useInterval. But then again when I saw the useEffect with an if and an interval and cleanup return func inside that if, I felt confused and a bit uncertain of how that works, and worse, if/how that could break.
So I decided to flip that and run the interval outside the if, and have the if check inside.

This means the interval now will run even if it's not a live auction and the code is simpler. I felt I'm ok with that tradeoff.

admbtlr commented on src/lib/Scenes/Sale/helpers/__tests__/auctionTime-tests.ts:
```diff
@@ -1,105 +1,227 @@
-const TIME_ZONE_NY = "America/New_York"
-const TIME_ZONE_LA = "America/Los_Angeles"
-const TIME_ZONE_DE = "Europe/Berlin"
+const timezones = {
```
Nice work on these tests!

admbtlr commented on src/lib/Scenes/Sale/helpers/saleTime.ts:
```diff
@@ -1,68 +1,122 @@
+    return closed(endDateMoment)
+  }
+
+  // if we have both start and end and we in the middle
```
```suggestion
  // if we have both start and end and we're in between them
```

pvinis replied:
🇬🇧

admbtlr commented on src/lib/utils/usePrevious.ts:
```diff
@@ -0,0 +1,11 @@
+import { useEffect, useRef } from "react"
```
Do we need this util? This is a known pattern, and hiding it in a 6 line library function is kinda obfuscating

pvinis replied:
🤔 I'm not sure I understand 😬. What do you mean?

The `usePrevious` is kind of a nice and helpful hook. Apparently we are not using it anywhere in eigen already, so I thought I'll add this here.

admbtlr replied:
Actually, you're right, creating a named hook to do this is a nice approach 👍 . (I was just thinking that it was unnecessary to create a separate hook in order to abstract away 6 lines of code)

pvinis replied:
Aaah gotcha. Thanks for explaining. Yes, I think it's nice to have, and convenient to have it as a reusable hook.

pvinis commented on src/lib/Scenes/Sale/Sale.tsx:
```diff
@@ -48,29 +51,45 @@ export const Sale: React.FC<Props> = ({ queryRes }) => {
-      }
+  // poll every .5 seconds to check if sale has gone live
+  useInterval(() => {
+    const now = moment()
+    if (sale.liveStartAt === null) {
```
```suggestion
    if (sale.liveStartAt === null) {
      const now = moment()
```
I'll quickly do this. This feels slightly nicer.

pvinis commented on src/lib/Scenes/Sale/Sale.tsx:
```diff
@@ -48,29 +51,45 @@ export const Sale: React.FC<Props> = ({ queryRes }) => {
-      }
+  // poll every .5 seconds to check if sale has gone live
+  useInterval(() => {
+    const now = moment()
+    if (sale.liveStartAt === null) {
+      setIsLive(false)
+      return
     }
```
```suggestion
    if (sale.liveStartAt === null) {
      setIsLive(false)
      return
    }
    const now = moment()
```
I'll do this. This feels slightly nicer.

MounirDhahri commented:
Thanks for adding the tests and the quick fixes 🌷
